File: code/Cpp/02_factory_method.cpp
/**
 * @file        - 02_factory_method.cpp
 * @author      - wdn (dongnian.wang@outlook.com)
 * @brief       - 工厂模式
 * @version     - 0.1
 * @date        - 2022-05-07
 * 
 * @copyright Copyright (c) 2022 
 * 
 */

#include <iostream>
#include <string>

/**
 *抽象基类
*/
class Document
{
public:
    Document(const std::string& name) : m_name(name) {}

    virtual void open() = 0;
    virtual void close() = 0;

    std::string get_name()
    {
        return m_name;
    }

private:
    std::string m_name;
};

/**
 * 具体派生类
*/
class MyDocument : public Document
{
public:
    MyDocument(const std::string& name) : Document(name) {}

    void open()
    {
        std::cout  << "   MyDocument: Open()" << std::endl;
    }

    void close()
    {
        std::cout  << "   MyDocument: Open()" << std::endl;
    }
};

class Application
{
public:
    Application() : m_index(0)
    {
        std::cout << "Application: ctor" << std::endl;
    }


    void new_document(const std::string& name)
    {
        std::cout << "Application: new document() " << std::endl;
        m_doc[m_index] = create_document(name);
        m_doc[m_index++]->open();
    }

    void open_document() {};

    void report_docs()
    {
        std::cout << "Application: report_docs()" << std::endl;
        for (int i = 0; i < m_index; i++)
            std::cout << "    " << m_doc[i]->get_name() << std::endl;
    }

    virtual Document* create_document(std::string name) = 0;

private:
    int m_index;
    Document* m_doc[10];
};

class MyApplication : public Application
{
public:
    MyApplication()
    {
        std::cout << "MyApplication: ctor" << std::endl;
    }

    Document* create_document(std::string name) override
    {
        std::cout << "   MyApplication: CreateDocument()" << std::endl;
        return new MyDocument(name);
    }
};

int main()
{
    MyApplication app;

    app.new_document("foo");
    app.new_document("bar");

    app.report_docs();

    return 0;
}

File: code/Python/02_factory_method.py
# -*- coding: utf-8 -*-
#  @file        - 02_factory_method.py
#  @author      - wdn (dongnian.wang@outlook.com)
#  @brief       - 工厂模式
#  @version     - 0.1
#  @date        - 2022-03-10
#  @copyright   - Copyright (c) 2021 


import abc

class AbstractFactory(metaclass=abc.ABCMeta):
    """
    抽象工厂类：提供产品的接口，调用者通过它访问具体工厂创建产品
    """
    
    def __init__(self) -> None:
        self.product = self._factory_method()
    
    @abc.abstractmethod
    def _factory_method(self):
        pass

    def some_operation(self):
        self.product.interface()

class ConcreteFactory1(AbstractFactory):
    """
    具体工厂1，实现抽象工厂中的抽象方法，创建产品
    """
    def _factory_method(self):
        print("Concrete Factory 1, create Concrete Product 1")
        return ConcreteProduct1()

class ConcreteFactory2(AbstractFactory):
    """
    具体工厂2，实现抽象工厂中的抽象方法，创建产品
    """

    def _factory_method(self):
        print("Concrete Factory 2, create Concrete Product 2")
        return ConcreteProduct2()
    
class AbstractProduct(metaclass=abc.ABCMeta):
    """
    抽象产品：定义产品的规范，描述产品的主要特性
    """

    @abc.abstractmethod
    def interface(self):
        pass

class ConcreteProduct1(AbstractProduct):
    """
    具体产品1：实现抽象产品所定义的接口，由具体工厂来创建
    """

    def interface(self):
        print("Concrete Product 1")

class ConcreteProduct2(AbstractProduct):
    """
    具体产品2：实现抽象产品所定义的接口，由具体工厂来创建
    """

    def interface(self):
        print("Concrete Product 2")

def main():
    concrete_creator = ConcreteFactory1()
    concrete_creator.product.interface()
    concrete_creator.some_operation()

if __name__ == "__main__":
    main()   
